flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,261 @@
+		a.Alerter.LogNonFlowWarning(ctx, telemetry.MaintenanceWait, mirror.MirrorName, fmt.Sprintf(
+			"Maintenance mode is still waiting for mirror to finish snapshot, mirror=%s, workflowId=%s, status=%s",
+			mirror.MirrorName, mirror.WorkflowId, mirrorStatus))
+	})()
+
+	defer shared.Interval(ctx, logEvery, func() {
+		slog.Info("[Maintenance] Waiting for mirror to finish snapshot",
+			"mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "status", mirrorStatus.String())
+	})()
+
+	slog.Info("Checking and waiting if mirror is snapshot", "mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "status",
+		mirrorStatus.String())
+	snapshotWaitSleepInterval := 10 * time.Second
+	for mirrorStatus == protos.FlowStatus_STATUS_SNAPSHOT || mirrorStatus == protos.FlowStatus_STATUS_SETUP {
+		time.Sleep(snapshotWaitSleepInterval)
+		activity.RecordHeartbeat(ctx, fmt.Sprintf("Waiting for mirror %s to finish snapshot", mirror.MirrorName))
+		mirrorStatus, err = a.getMirrorStatus(ctx, mirror)
+		if err != nil {
+			return mirrorStatus, err
+		}
+	}
+	return mirrorStatus, nil
+}
+
+func (a *MaintenanceActivity) EnableMaintenanceMode(ctx context.Context) error {
+	slog.Info("Enabling maintenance mode")
+	return peerdbenv.UpdatePeerDBMaintenanceModeEnabled(ctx, a.CatalogPool, true)
+}
+
+func (a *MaintenanceActivity) BackupAllPreviouslyRunningFlows(ctx context.Context, mirrors *protos.MaintenanceMirrors) error {
+	tx, err := a.CatalogPool.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	defer shared.RollbackTx(tx, slog.Default())
+
+	for _, mirror := range mirrors.Mirrors {
+		_, err := tx.Exec(ctx, `
+		insert into maintenance.maintenance_flows
+			(flow_id, flow_name, workflow_id, flow_created_at, is_cdc, state, from_version)
+		values
+			($1, $2, $3, $4, $5, $6, $7)
+		`, mirror.MirrorId, mirror.MirrorName, mirror.WorkflowId, mirror.MirrorCreatedAt, mirror.IsCdc, mirrorStateBackup,
+			peerdbenv.PeerDBVersionShaShort())
+		if err != nil {
+			return err
+		}
+	}
+	return tx.Commit(ctx)
+}
+
+func (a *MaintenanceActivity) PauseMirrorIfRunning(ctx context.Context, mirror *protos.MaintenanceMirror) (bool, error) {
+	mirrorStatus, err := a.getMirrorStatus(ctx, mirror)
+	if err != nil {
+		return false, err
+	}
+
+	slog.Info("Checking if mirror is running", "mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "status", mirrorStatus.String())
+
+	if mirrorStatus != protos.FlowStatus_STATUS_RUNNING {
+		return false, nil
+	}
+
+	slog.Info("Pausing mirror for maintenance", "mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId)
+
+	err = model.FlowSignal.SignalClientWorkflow(ctx, a.TemporalClient, mirror.WorkflowId, "", model.PauseSignal)
+	if err != nil {
+		slog.Error("Error signaling mirror running to pause for maintenance",
+			"mirror", mirror.MirrorName, "workflowId", mirror.WorkflowId, "error", err)
+		return false, err
+	}
+	defer shared.Interval(ctx, 30*time.Second, func() {
+		slog.Info("Waiting for mirror to pause",
```

Comment:
can move this into loop alongside heartbeat

Reply:
wanted to have a different interval for this, don't need to spam logs for shorter runs

Reply:
just switching to a ticker here as well